src/main/java/org/manifold/compiler/front/ExpressionGraph.java

```diff
@@ -52,17 +49,19 @@ public void addVertex(VariableIdentifier vID)
   }
   public void addVertex(ExpressionVertex v) {
     nonVariableVertices.add(v);
+    allVertices.add(v);
   }
 
-  public List<ExpressionVertex> getVertices() {
-    List<ExpressionVertex> vertices = new LinkedList<>();
-    vertices.addAll(getVariableVertices().values());
-    vertices.addAll(getNonVariableVertices());
-    return vertices;
+  public Collection<ExpressionVertex> getVertices() {
+    return allVertices;
   }
 
   private List<ExpressionEdge> edges = new ArrayList<>();
   public void addEdge(ExpressionEdge e) {
+    Preconditions.checkArgument(
+        getVertices().contains(e.getSource())
+            && e.getTarget() == null || getVertices().contains(e.getTarget()),
```

Comment:
`A && B || C` is parsed left to right as `(A && B) || C`. I think you mean 
` && (e.getTarget() == null || getVertices().contains(e.getTarget()))`